jmclient/jmclient/wallet_rpc.py

```diff
@@ -0,0 +1,674 @@
+            #when no wallet status is false
+            session = not self.cookie==None
+            maker_running = False
+            if self.services["maker"] and self.services["maker"].isRunning():
+                maker_running = True
+            if self.taker:
+                coinjoin_in_process = True
+            else:
+                coinjoin_in_process = False
+            if self.wallet_service:
+                if self.wallet_service.isRunning():
+                    wallet_name = self.wallet_name
+                else:
+                    wallet_name = "not yet loaded"
+            else:
+                wallet_name = "None"
+            return make_jmwalletd_response(request,session=session,
+                            maker_running=maker_running,
+                            coinjoin_in_process=coinjoin_in_process,
+                            wallet_name=wallet_name)
+
+        @app.route('/wallet/<string:walletname>/snicker/start', methods=['GET'])
+        def start_snicker(self, request, walletname):
+            self.check_cookie(request)
+            if not self.wallet_service:
+                raise NoWalletFound()
+            if self.services["snicker"] and self.services["snicker"].isRunning():
+                raise ServiceAlreadyStarted()
+            # TODO: allow client to inject acceptance callbacks to Receiver
+            self.services["snicker"] = SNICKERReceiverService(
+                SNICKERReceiver(self.wallet_service))
+            self.services["snicker"].startService()
+            # TODO waiting for startup seems perhaps not needed here?
+            return make_jmwalletd_response(request, walletname=walletname)
+
+        @app.route('/wallet/<string:walletname>/snicker/stop', methods=['GET'])
+        def stop_snicker(self, request, walletname):
+            self.check_cookie(request)
+            if not self.wallet_service:
+                raise NoWalletFound()
+            if not self.services["snicker"]:
+                raise ServiceNotStarted()
+            self.services["snicker"].stopService()
+            return make_jmwalletd_response(request, walletname=walletname)
+
+        @app.route('/wallet/<string:walletname>/taker/direct-send', methods=['POST'])
+        def directsend(self, request, walletname):
+            """ Use the contents of the POST body to do a direct send from
+            the active wallet at the chosen mixdepth.
+            """
+            self.check_cookie(request)
+            assert isinstance(request.content, BytesIO)
+            payment_info_json = self.get_POST_body(request, ["mixdepth", "amount_sats",
+                                                             "destination"])
+            if not payment_info_json:
+                raise InvalidRequestFormat()
+            if not self.wallet_service:
+                raise NoWalletFound()
+            tx = direct_send(self.wallet_service, int(payment_info_json["amount_sats"]),
```

Comment:
OK, so to focus specifically on direct send for a moment: reviewing the existing code I see we need to (a) change some `None` returns to `False` in the function, (b) while Qt handles false-y returns and exceptions as expected, sendpayment currently just allows assertion error crashes and relies on being able to read logged errors and stack traces (not ideal but not a disaster in itself).

So I propose to make these changes: in directsend,  make sure always to return False as according to the docstring, when needed, and catch all Exceptions with error messages (obviously mainly the Assertion errors, but there may be others) here in our directsend API call and return it as a `400 BadRequest` based on the idea that even if the HTTP request was syntactically well formed, the wallet was being asked to do something it can't do; or is that an abuse of code 400? Does 412 make sense or is that  headers-specific somehow?

Any similar issues with docoinjoin, I'll fold into the discussion of that, below.